File: implement-shell-tools/cat/mycat.js
import { program } from "commander";
import { promises as fs } from "node:fs";
import process from "node:process";

program
  .name("mycat")
  .description("Outputs the content of the given file(s), like the cat command")
  .option("-n", "Number all lines")
  .option("-b", "Number non-blank lines only")
  .argument("<files...>", "File paths to display");

program.parse();

const options = program.opts();
const filePaths = program.args;

let lineNumber = 1;

for (const path of filePaths) {
  try {
    const content = await fs.readFile(path, "utf-8");
    const lines = content.trimEnd().split("\n");

    for (const line of lines) {
      const isBlank = line.trim() === "";

      if (options.b) {
        // -b: number only non-blank lines
        if (!isBlank) {
          process.stdout.write(`${String(lineNumber).padStart(6)}  ${line}\n`);
          lineNumber++;
        } else {
          process.stdout.write("\n");
        }
      } else if (options.n) {
        // -n: number all lines
        process.stdout.write(`${String(lineNumber).padStart(6)}  ${line}\n`);
        lineNumber++;
      } else {
        // no flags
        process.stdout.write(line + "\n");
      }
    }
  } catch (error) {
    process.stderr.write(`Error reading file ${path}: ${error.message}\n`);
    process.exit(1);
  }
}

File: implement-shell-tools/cat/my_cat.py
import argparse

parser = argparse.ArgumentParser(
    prog="my_cat",
    description="The re-implementation of the cat command",
    epilog="Example usage: my_cat file1.txt file2.txt -n"
)

parser.add_argument("files", nargs="+", help="files to read")
parser.add_argument("-n", "--number", action="store_true", help="number all lines")
parser.add_argument("-b", "--number-nonblank", action="store_true", help="number non-empty lines")

args = parser.parse_args()
print(args)
line_number = 1

try:
    for path in args.files:
        with open(path, "r", encoding="utf-8") as f:
            lines = f.readlines()

        if args.number_nonblank:
            output = []
            for line in lines:
                if line.strip() == "":
                    output.append(line.rstrip("\n"))
                else:
                    output.append(f"{line_number} {line.rstrip()}")
                    line_number += 1

        elif args.number:
            output = [
                f"{i + 1} {line.rstrip()}"
                for i, line in enumerate(lines)
            ]

        else:
            output = [line.rstrip("\n") for line in lines]

        print("\n".join(output))

except Exception as err:
    print(f"Error reading file: {err}")
    exit(1)